### DISCUSSION TITLE
Support private repositories for pipx/uvx

### DISCUSSION THREAD

**Discussion by clement-tourriere** [jdx/mise#3893] [Ideas]
For the moment Mise only supports package installation from PYPI (and github/git url).

Could you add the possibility to support package installation from custom python package index url via custom --index-url option for [pipx](https://pipx.pypa.io/stable/docs/#pipx-install) and --default-index parameter for uv(x) or even by defining a [UV_DEFAULT_INDEX](https://docs.astral.sh/uv/configuration/environment/#uv_default_index) variable in`mise.config.toml`

**Comment by jdx** [comment]
I think there's something like `pipx_install_args` and `install_env` (check docs) which can be used to set either of these

**Reply by clement-tourriere** [reply]
Hello thank you for your very quick answer and I'm really sorry for mine, I was in holidays.

I tried to follow the documentation and add the parameters before creating this issue, but I can't make it work. `Mise` always tries to retrieve package from pypi, whatever the options I try.

My config file looks like:

```
[settings]

[settings.pipx]
uvx = true

[tools]
"pipx:my_private_package" = { version = "1.2.3", uvx_args = "--default-index=https://my_full_package_url" }
```

Then, when I try to do the command `mise sync python`,  I still have this error: `HTTP status client error (404 Not Found) for url (https://pypi.org/pypi/<package>/json)` 

As a sidenote, I try to install a "global" tool, and it would really be practical to be able to pass an private index url when performing a global installation:

`mise use -g pipx:my_package --uvx-args=` or something like that

**Reply by jdx** [reply]
yeah as I explained in https://github.com/jdx/mise/discussions/4084 mise doesn't use that option (and probably shouldn't) for listing versions, so you can either express full versions or we'll need a new option to handle this

**Reply by clement-tourriere** [reply]
Thank you for your reply, but I'm not sure to understand. I also read the code before, and since you are performing a call to pypi.org to check the package version, it cannot work for a private package that is not present on pypi.org at all. I think this use case can be quite common, to have a private repository with private packages that don't appear on pypi.org
But maybe there is something I don't understand, and if this is the case, could you provide a "working" example (or what I need to change in my config file to make it work)

**Reply by jdx** [reply]
I'm saying mise needs a new feature to handle this use-case

**Reply by anthonydurot** [reply]
Hi @jdx, what's the status on this please ? We would really like to have this feature.

Anyway, thanks for your awesome work !

**Reply by jdx** [reply]
I have no plans to work on this

**Reply by anthonydurot** [reply]
I think the issue comes from https://github.com/jdx/mise/blob/d35046fdd976bb412c88dade243e33842ebc32ab/src/backend/pipx.rs#L70

The fetching of versions doesn't include authentication headers, so it can't list versions
